hazelcast-sql/src/test/java/com/hazelcast/sql/impl/expression/ExpressionTestBase.java

```diff
@@ -0,0 +1,1190 @@
+import org.apache.calcite.sql.SqlOperandCountRange;
+import org.apache.calcite.sql.SqlOperator;
+import org.apache.calcite.sql.SqlSelect;
+import org.apache.calcite.sql.SqlSyntax;
+import org.apache.calcite.sql.parser.SqlParserPos;
+import org.apache.calcite.sql.type.SqlOperandCountRanges;
+import org.apache.calcite.sql.type.SqlTypeName;
+import org.apache.calcite.sql.validate.SqlValidator;
+import org.apache.calcite.sql.validate.SqlValidatorScope;
+import org.junit.After;
+
+import java.io.PrintWriter;
+import java.io.Serializable;
+import java.io.StringWriter;
+import java.math.BigDecimal;
+import java.math.BigInteger;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+import java.util.function.Function;
+
+import static com.hazelcast.sql.impl.calcite.SqlToQueryType.mapRowType;
+import static com.hazelcast.sql.impl.calcite.validate.types.HazelcastTypeSystem.narrowestTypeFor;
+import static com.hazelcast.sql.impl.calcite.validate.types.HazelcastTypeSystem.withHigherPrecedence;
+import static com.hazelcast.sql.impl.calcite.validate.types.HazelcastTypeSystem.withHigherPrecedenceForLiterals;
+import static java.util.Collections.singletonMap;
+import static org.apache.calcite.sql.type.SqlTypeName.ANY;
+import static org.apache.calcite.sql.type.SqlTypeName.BIGINT;
+import static org.apache.calcite.sql.type.SqlTypeName.BOOLEAN;
+import static org.apache.calcite.sql.type.SqlTypeName.CHAR_TYPES;
+import static org.apache.calcite.sql.type.SqlTypeName.DECIMAL;
+import static org.apache.calcite.sql.type.SqlTypeName.DOUBLE;
+import static org.apache.calcite.sql.type.SqlTypeName.FRACTIONAL_TYPES;
+import static org.apache.calcite.sql.type.SqlTypeName.INTEGER;
+import static org.apache.calcite.sql.type.SqlTypeName.INT_TYPES;
+import static org.apache.calcite.sql.type.SqlTypeName.NULL;
+import static org.apache.calcite.sql.type.SqlTypeName.NUMERIC_TYPES;
+import static org.apache.calcite.sql.type.SqlTypeName.REAL;
+import static org.apache.calcite.sql.type.SqlTypeName.SMALLINT;
+import static org.apache.calcite.sql.type.SqlTypeName.TIMESTAMP_WITH_LOCAL_TIME_ZONE;
+import static org.apache.calcite.sql.type.SqlTypeName.TINYINT;
+import static org.apache.calcite.sql.type.SqlTypeName.VARCHAR;
+import static org.junit.Assert.assertArrayEquals;
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertNull;
+import static org.junit.Assert.assertSame;
+import static org.junit.Assert.assertTrue;
+import static org.junit.Assert.fail;
+
+public abstract class ExpressionTestBase extends SqlTestSupport {
```

Comment:
The chosen approach with generated cases is extremely difficult to follow. I cannot reason on whether the coverage is sufficient or not, because looking at thousands of invocations requires too much time. Neither I can understand how to deal with it in the future - what should I do as a developer when a new expression is introduced, or when a failure is observed? And how do I ensure that the test itself is correct when it is filled with complicated logic?

I would propose two things:
1) We need detailed documentation about the testing framework, so that other developers know how to deal with it
2) Given that the complexity of the framework is high, we'd better rely on a set of simple manually written end-to-end tests to get quick understanding of the behavior in common cases, and use this framework to cover the edge cases

Reply:
Yes, I'm also not that happy about it. But the "framework" allowed to catch numerous bugs and edge cases while working on the expressions. Initially, I started adding manually written tests for every expression, but then it became clear no matter how many tests I'm adding where is always one more to add covering some yet undiscovered case, so I switched to exhaustive enumeration. Think of it as a tool to verify the integration with Calcite, not as a fully fledged SQL test suite replacing end-to-end tests. Unfortunately, we (well, basically me) don't have enough capacity to work on good quality test suite and Calcite integration in parallel.

Reply:
Regarding the documentation, I will add it. Usually when a new expression is added, an empty test case is created with expected types/values lambdas always returning "invalid" value, then the missing logic is created on all three sides: expression implementation itself, Calcite integration and in the test case, until the test passes.

Reply:
Ok, I see - there are no easy answers here. We should have a clear view on the purpose of these tests then - what they test, and what they do not test. AFAIU from your comment, we do use them mainly to test Calcite integration. But we cannot use them to test public behavior, such as returned values, column metadata, exception messages. 

Therefore, we need to cover basic tests manually anyway, and for the most expressions, the manual coverage is not that sufficient at the moment, especially around exceptions and metadata. 

Though, I think this framework could be a good way to understand what to test manually. E.g by looking at generated tests we can see some erroneous cases with specific errors and add manual tests for them.